### DISCUSSION TITLE
Allow setting `onlyBuiltDependencies` globally

### DISCUSSION THREAD

**Discussion by benmccann** [pnpm/pnpm#15795] [Ideas]
### Contribution

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

A vulnerable version of rspack was just published, which had a malicious `postinstall` script. Add new configuration values that would project against this without creating too much frustration for users. Most people don't want to disable postinstall scripts because some libraries expect them to be used. However, most use cases for postinstall scripts are for a small set of known libraries or for the user's own project rather than for dependencies

### Describe the solution you'd like

`enable-pre-post-scripts` could have a value such as `'project'`, which would run only the project's `preinstall` and `postinstall` scripts and not those of any dependencies

Additionally, `enable-pre-post-scripts` could accept an array of whitelisted dependencies that would be allowed to run `preinstall` and `postinstall` scripts

### Describe the drawbacks of your solution

There are few downsides to introducing a new option. If one of these options were made the default that would introduce friction for some users, but I believe the ecosystem would be made better off and that trade-off may be worth it

### Describe alternatives you've considered

Only do JavaScript development inside VMs or containers. Realistically, most users aren't going to do this.

**Comment by zkochan** [comment]
Well, first of all you have the wrong idea about the `enable-pre-post-scripts` setting. This setting is not about disabling postinstall but about disabling pre<scriptName> and post<scriptName>, when you run `pnpm run <scriptName>`.

If you want to disallow build scripts of dependencies, you can use [pnpm.onlyBuiltDependencies](https://pnpm.io/package_json#pnpmonlybuiltdependencies) to list only those dependencies that are allowed to be built. So, if a dependency is not in that list, the postinstall of that dependency will be ignored.

**Comment by zkochan** [comment]
I don't think this is needed as we want to block lifecycle scripts by default in v10. Related PR: https://github.com/pnpm/pnpm/pull/8897

**Comment by benmccann** [comment]
Thank you. Blocking lifecycle scripts by default is exactly what I was looking for

**Comment by benmccann** [comment]
Reopening this as the implemented solution required every project to be updated. I would like to be able to say that `esbuild` should always be allowed to run its postinstall script

**Comment by zkochan** [comment]
If you will do it globally on your local machine then it won't work on someone else's machine.

You can create a package with a JSON list of package names that are allowed to be built. You can then install that package as a [configurational dependency](https://pnpm.io/config-dependencies) and in your package add a reference to that file in the dependency via [pnpm.onlyBuiltDependenciesFile](https://pnpm.io/package_json#pnpmonlybuiltdependenciesfile)

It will look like something like this:

```json
{
  "pnpm": {
    "configDependencies": {
      "my-allow-list-pkg": "1.0.0+sha512-30iZtAPgz+LTIYoeivqYo853f02jBYSd5uGnGpkFV0M3xOt9aN73erkgYAmZU43x4VfqcnLxW9Kpg3R5LC4YYw=="
    },
    "onlyBuiltDependenciesFile": "node_modules/.pnpm-config/my-allow-list-pkg/onlyBuiltDependencies.json"
  }
}
```

**Comment by zkochan** [comment]
Alternatively, if you want to go back to enabling all scripts, you can add an empty neverBuiltDependencies setting to the package.json:

```json
{
  "pnpm": {
    "neverBuiltDependencies": []
  }
}
```

But that's probably not a good idea.

**Comment by karlhorky** [comment]
@zkochan is a global `onlyBuiltDependencies` setting relevant for approving build scripts for packages to be globally installed with `pnpm add --global`?

Eg. what is the recommendation for building `esbuild` in this case? Should the user be modifying the `/Users/k/Library/pnpm/global/5/package.json` file with `pnpm.onlyBuiltDependencies = ['esbuild']`?

```bash
$ pnpm add --global @upleveled/preflight
Packages: +132
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 781, reused 758, downloaded 0, added 0, done

/Users/k/Library/pnpm/global/5:
+ @upleveled/preflight 8.0.4

╭ Warning ───────────────────────────────────────────────────────────────────────────────────╮
│                                                                                            │
│   Ignored build scripts: esbuild.                                                          │
│   Run "pnpm approve-builds" to pick which dependencies should be allowed to run scripts.   │
│                                                                                            │
╰────────────────────────────────────────────────────────────────────────────────────────────╯
```

**Comment by karlhorky** [comment]
## Workaround (programmatically add to global `pnpm.onlyBuiltDependencies`)

We're currently working around setting `pnpm.onlyBuiltDependencies` globally for `pnpm add --global` on CI for now by editing the `package.json` in the pnpm global root (eg. `/Users/k/Library/pnpm/global/5/package.json`):

```ts
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { execa } from 'execa';

// Workaround to globally set `pnpm.onlyBuiltDependencies` for `esbuild`
// - https://github.com/pnpm/pnpm/issues/8891#issuecomment-2651840685
const globalPnpmRoot = (await execa`pnpm root --global`).stdout.replace(
  /[/\\]node_modules$/,
  '',
);
await mkdir(globalPnpmRoot, { recursive: true });
const globalPnpmPackageJsonPath = `${globalPnpmRoot}/package.json`;

let globalPnpmPackageJson: {
  pnpm?: { onlyBuiltDependencies?: string[] };
} = {};

try {
  globalPnpmPackageJson = JSON.parse(
    await readFile(globalPnpmPackageJsonPath, 'utf8'),
  );
} catch {
  // Swallow error if package.json doesn't exist
}

if (
  !globalPnpmPackageJson.pnpm?.onlyBuiltDependencies ||
  !globalPnpmPackageJson.pnpm.onlyBuiltDependencies.includes('esbuild')
) {
  console.log(
    `Updating pnpm global package.json at ${globalPnpmPackageJsonPath} to add 'esbuild' to pnpm.onlyBuiltDependencies...`,
  );
  globalPnpmPackageJson.pnpm = globalPnpmPackageJson.pnpm || {};
  globalPnpmPackageJson.pnpm.onlyBuiltDependencies =
    globalPnpmPackageJson.pnpm.onlyBuiltDependencies || [];
  globalPnpmPackageJson.pnpm.onlyBuiltDependencies.push('esbuild');
  await writeFile(
    globalPnpmPackageJsonPath,
    JSON.stringify(globalPnpmPackageJson, null, 2) + '\n',
  );
}
```

Feels like this is not a great solution though... 🤔 Haven't found any other CLI commands that would allow us to achieve this though yet.

**Comment by zkochan** [comment]
IMO, for the global packages we need to do it similarly to how we do it for `pnpm dlx`, via a flag. So, in your example it would be done like this:

```
pnpm add --global @upleveled/preflight --allow-build=esbuild
```

**Comment by zkochan** [comment]
Related PR: https://github.com/pnpm/pnpm/pull/9086

**Comment by karlhorky** [comment]
> If you will do it globally on your local machine then it won't work on someone else's machine.

@zkochan after the Axios breach today, I'm thinking it's still good to configure ALSO globally for defense in depth, eg:

1. installing global dependencies
2. installing in projects which don't yet have `pnpm-workspace.yaml` config
